### ISSUE TITLE
MdnsEvent::Expired (ExpiredAddrsIter) is never called 

### ISSUE THREAD

**Issue by MariaGolubev** [libp2p/rust-libp2p#2308]
I wrote a small program. I expected that when a node exits, other nodes will receive the MdnsEvent :: Expired event, but this does not happen. I looked at the code for a long time, and I realized that the problem is: 
```Rust
if let Some(ref mut closest_expiration) = self.closest_expiration {
            if let Poll::Ready(now) = Pin::new(closest_expiration).poll(cx) {
                let mut expired = SmallVec::<[(PeerId, Multiaddr); 4]>::new();
                while let Some(pos) = self
                    .discovered_nodes
                    .iter()
                    .position(|(_, _, exp)| *exp < now)
                {
                    let (peer_id, addr, _) = self.discovered_nodes.remove(pos);
                    expired.push((peer_id, addr));
                }
                
                if !expired.is_empty() {
                    let event = MdnsEvent::Expired(ExpiredAddrsIter {
                        inner: expired.into_iter(),
                    });

                    return Poll::Ready(NetworkBehaviourAction::GenerateEvent(event));
                }
            }
        }
```
i wrapped expired in dbg!
```Rust
dbg!(expired);
```
it turned out that it is always empty 
```sh
[/home/golubevmt/.cargo/registry/src/github.com-1ecc6299db9ec823/libp2p-mdns-0.31.0/src/behaviour.rs:369] &expired = []
[/home/golubevmt/.cargo/registry/src/github.com-1ecc6299db9ec823/libp2p-mdns-0.31.0/src/behaviour.rs:369] &expired = []
[/home/golubevmt/.cargo/registry/src/github.com-1ecc6299db9ec823/libp2p-mdns-0.31.0/src/behaviour.rs:369] &expired = []
[/home/golubevmt/.cargo/registry/src/github.com-1ecc6299db9ec823/libp2p-mdns-0.31.0/src/behaviour.rs:369] &expired = []
```
i changed `
```Rust 
*exp < now
````
 to 
```Rust 
ext  <= &now
```
and it worked for me 

```sh
Oct 23 00:47:04.706  INFO p2p_test: Discovered: 12D3KooWSK4J25g3WNFWo8n7DRxNK5Rwz3dsgDdRKHwowDgr1cbR /ip4/192.168.199.16/udp/57229/quic
Oct 23 00:47:04.706  INFO p2p_test: Discovered: 12D3KooWSK4J25g3WNFWo8n7DRxNK5Rwz3dsgDdRKHwowDgr1cbR /ip4/127.0.0.1/udp/57229/quic
[/home/golubevmt/.cargo/registry/src/github.com-1ecc6299db9ec823/libp2p-mdns-0.31.0/src/behaviour.rs:364] exp <= &now = true
[/home/golubevmt/.cargo/registry/src/github.com-1ecc6299db9ec823/libp2p-mdns-0.31.0/src/behaviour.rs:364] exp <= &now = true
Oct 23 00:47:10.711  INFO p2p_test: Expired: 12D3KooWSK4J25g3WNFWo8n7DRxNK5Rwz3dsgDdRKHwowDgr1cbR /ip4/192.168.199.16/udp/57229/quic
Oct 23 00:47:10.712  INFO p2p_test: Expired: 12D3KooWSK4J25g3WNFWo8n7DRxNK5Rwz3dsgDdRKHwowDgr1cbR /ip4/127.0.0.1/udp/57229/quic
```

**Comment by thomaseizinger** [comment]
Thanks!

It would be awesome if you could submit a PR for this including a test :)

**Comment by MariaGolubev** [comment]
I added fixes, but two tests still fail: 
```sh
test test_expired_async_std_ipv4 ... ok
test test_expired_tokio_ipv4 ... ok
test test_expired_tokio_ipv6 ... FAILED
test test_expired_async_std_ipv6 ... FAILED
```
I don’t know the reason. I have no problem in version `0.39`, but in `0.40.0-rc.2` the expired `IPV6_MDNS_MULTICAST_ADDRESS` is not called. I don't have time to study the problem, please take it

**Comment by MariaGolubev** [comment]
# here are the tests
```Rust

async fn run_expired_test(config: MdnsConfig) -> Result<(), Box<dyn Error>> {
    let mut a = create_swarm(config.clone()).await?;
    let mut b = create_swarm(config).await?;
    loop {
        futures::select! {
            ev = a.select_next_some() => match ev {
                SwarmEvent::Behaviour(MdnsEvent::Expired(peers)) => {
                    for (peer, _addr) in peers {
                        if peer == *b.local_peer_id() {
                            return Ok(());
                        }
                    }
                }
                _ => {}
            },
            ev = b.select_next_some() => match ev {
                SwarmEvent::Behaviour(MdnsEvent::Expired(peers)) => {
                    for (peer, _addr) in peers {
                        if peer == *a.local_peer_id() {
                            return Ok(());
                        }
                    }
                }
                _ => {}
            }

        }
    }
}

#[async_std::test]
async fn test_expired_async_std_ipv4() -> Result<(), Box<dyn Error>> {
    let config = MdnsConfig {
        ttl: Duration::from_millis(500),
        query_interval: Duration::from_secs(1),
        ..Default::default()
    };

    async_std::future::timeout(Duration::from_secs(3), run_expired_test(config))
        .await
        .unwrap()
}

#[async_std::test]
async fn test_expired_async_std_ipv6() -> Result<(), Box<dyn Error>> {
    let config = MdnsConfig {
        ttl: Duration::from_millis(500),
        query_interval: Duration::from_secs(1),
        multicast_addr: *IPV6_MDNS_MULTICAST_ADDRESS,
    };

    async_std::future::timeout(Duration::from_secs(3), run_expired_test(config))
        .await
        .unwrap()
}

#[tokio::test]
async fn test_expired_tokio_ipv4() -> Result<(), Box<dyn Error>> {
    let config = MdnsConfig {
        ttl: Duration::from_millis(500),
        query_interval: Duration::from_secs(1),
        ..Default::default()
    };

    tokio::time::timeout(Duration::from_secs(3), run_expired_test(config))
        .await
        .unwrap()
}

#[tokio::test]
async fn test_expired_tokio_ipv6() -> Result<(), Box<dyn Error>> {
    let config = MdnsConfig {
        ttl: Duration::from_millis(500),
        query_interval: Duration::from_secs(1),
        multicast_addr: *IPV6_MDNS_MULTICAST_ADDRESS,
    };

    tokio::time::timeout(Duration::from_secs(3), run_expired_test(config))
        .await
        .unwrap()
}

```

**Comment by vnermolaev** [comment]
are there any updates? Is it being looked into?

**Comment by mxinden** [comment]
I don't have the capacity to look into this. Help greatly appreciated @vnermolaev.

**Comment by vnermolaev** [comment]
I started looking into this and it appears the tests in `smoke.rs` are faked, perhaps unintentionally, for IPv6.
Observe a test [here](https://github.com/libp2p/rust-libp2p/blob/master/protocols/mdns/tests/smoke.rs#L85).
For the sake of completeness, here it is:

```rust
#[async_std::test]
async fn test_discovery_async_std_ipv6() -> Result<(), Box<dyn Error>> {
    let mut config = MdnsConfig::default();
    config.multicast_addr = *IPV6_MDNS_MULTICAST_ADDRESS;
    run_test(MdnsConfig::default()).await
}
```

Although a config requiring IPv6 is created, it is never used and a default version with IPv4 is called. Passing the variable `config` breaks the test, so there is something broken deeper in the protocol. In fact, [line 36](https://github.com/libp2p/rust-libp2p/blob/master/protocols/mdns/tests/smoke.rs#L36) always instructs the swarm to listen on an ip4 address. Unfortunately, fixing it to listen to either ip4 or ip6 loopback dependently on the config does not help.

So as far as I am concerned, the fix proposed by @RustPanda is very much valid and only fails because peer discovery fails for IPv6, too.
I can try to look deeper, but, I beleive, you are much more familiar with the code.

**Comment by vnermolaev** [comment]
@mxinden any comments?

**Comment by vnermolaev** [comment]
Further details:
I did some further digging. There is a multiplicity of reasons why IPv6 might not work: 
- OS is not being configured to support it,
- Network DHCP is not configured to assign one,
- Impossibility to join the multicast address,
- ... and probably more.

And the bad part of it is that there is no error-handling (just `log::error`) for all of these cases. That is, while PR #2161 by @dvc94ch technically enabled IPv6, it has cut some corners (see example in my previous message).

To summarize, I believe, the solution given by @RustPanda (and tests) is valid in the same way PR #2161 was valid. Shall I prepare a PR that bypasses direct IPv6 testing in the same way PR #2161 does? @mxinden

**Comment by dvc94ch** [comment]
what's wrong with the ipv6 tests? looks like it should be `<=` instead of `<`, but I don't see what that has to do with ipv6.

**Comment by dvc94ch** [comment]
and what does the listening address have to do with anything?

**Comment by dvc94ch** [comment]
> I did some further digging. There is a multiplicity of reasons why IPv6 might not work:

> OS is not being configured to support it,
Network DHCP is not configured to assign one,
Impossibility to join the multicast address,
... and probably more.

> And the bad part of it is that there is no error-handling (just log::error) for all of these cases.

Not sure what you want to do. reconfigure the OS and network, patch the router firmware with ipv6 support? If you have a reasonable way to recover from one of those I welcome suggestions.

**Comment by dvc94ch** [comment]
so if the tests pass in ci it's fine, I assume the OP's machine is missconfigured and there is no ipv6 support.

**Comment by vnermolaev** [comment]
> what's wrong with the ipv6 tests? looks like it should be <= instead of <, but I don't see what that has to do with ipv6.

@dvc94ch It is indeed the case. The OP is testing peer disconnection for both ipv4 and ipv6 _faithfully_. While the tests for peer discovery do not seem to do so.

Please consider [one of the tests for ipv6](https://github.com/libp2p/rust-libp2p/blob/master/protocols/mdns/tests/smoke.rs#L85) submitted as part of PR #2161.
Correct me if I am wrong, but the created `config` variable instructing to use IPv6 is never passed to the test, instead `MdnsConfig::default()` is always passed, and the latter instructs to use IPv4.

So the problem is that it is impossible to distinguish why the tests for ipv6 fail: is it because the ipv6 is not supported or because the peer has not been judged as disconnected.

**Comment by dvc94ch** [comment]
ah yes, that seems to be a typo

**Comment by dvc94ch** [comment]
didn't actually write the code, but found the issue. someone passed an array of ips to UdpSocket::bind, don't know what genious mind that originated from.

**Comment by dvc94ch** [comment]
so heres the diff in case someone wants to open a PR:
```diff
diff --git a/protocols/mdns/Cargo.toml b/protocols/mdns/Cargo.toml
index 43370a9b..0cfd16c9 100644
--- a/protocols/mdns/Cargo.toml
+++ b/protocols/mdns/Cargo.toml
@@ -26,5 +26,6 @@ void = "1.0.2"
 
 [dev-dependencies]
 async-std = { version = "1.9.0", features = ["attributes"] }
+env_logger = "0.9.0"
 libp2p = { path = "../.." }
 tokio = { version = "1.2.0", default-features = false, features = ["macros", "rt", "rt-multi-thread"] }
diff --git a/protocols/mdns/src/behaviour.rs b/protocols/mdns/src/behaviour.rs
index a3c90aeb..7a8a09ab 100644
--- a/protocols/mdns/src/behaviour.rs
+++ b/protocols/mdns/src/behaviour.rs
@@ -144,12 +144,12 @@ impl Mdns {
             }
         };
         let send_socket = {
-            let addrs = [
-                SocketAddr::new(IpAddr::V4(Ipv4Addr::UNSPECIFIED), 0),
-                SocketAddr::new(IpAddr::V6(Ipv6Addr::UNSPECIFIED), 0),
-            ];
+            let addr = match config.multicast_addr {
+                IpAddr::V4(_) => SocketAddr::new(IpAddr::V4(Ipv4Addr::UNSPECIFIED), 0),
+                IpAddr::V6(_) => SocketAddr::new(IpAddr::V6(Ipv6Addr::UNSPECIFIED), 0),
+            };
 
-            let socket = std::net::UdpSocket::bind(&addrs[..])?;
+            let socket = std::net::UdpSocket::bind(addr)?;
             Async::new(socket)?
         };
         let if_watch = if_watch::IfWatcher::new().await?;
@@ -406,7 +406,7 @@ impl NetworkBehaviour for Mdns {
                 while let Some(pos) = self
                     .discovered_nodes
                     .iter()
-                    .position(|(_, _, exp)| *exp < now)
+                    .position(|(_, _, exp)| *exp <= now)
                 {
                     let (peer_id, addr, _) = self.discovered_nodes.remove(pos);
                     expired.push((peer_id, addr));
diff --git a/protocols/mdns/tests/smoke.rs b/protocols/mdns/tests/smoke.rs
index 3d374dd4..63671ff0 100644
--- a/protocols/mdns/tests/smoke.rs
+++ b/protocols/mdns/tests/smoke.rs
@@ -28,6 +28,7 @@ use libp2p::{
 use std::error::Error;
 
 async fn create_swarm(config: MdnsConfig) -> Result<Swarm<Mdns>, Box<dyn Error>> {
+    env_logger::try_init().ok();
     let id_keys = identity::Keypair::generate_ed25519();
     let peer_id = PeerId::from(id_keys.public());
     let transport = libp2p::development_transport(id_keys).await?;
@@ -85,7 +86,7 @@ async fn test_discovery_async_std_ipv4() -> Result<(), Box<dyn Error>> {
 async fn test_discovery_async_std_ipv6() -> Result<(), Box<dyn Error>> {
     let mut config = MdnsConfig::default();
     config.multicast_addr = *IPV6_MDNS_MULTICAST_ADDRESS;
-    run_test(MdnsConfig::default()).await
+    run_test(config).await
 }
 
 #[tokio::test]
@@ -97,5 +98,5 @@ async fn test_discovery_tokio_ipv4() -> Result<(), Box<dyn Error>> {
 async fn test_discovery_tokio_ipv6() -> Result<(), Box<dyn Error>> {
     let mut config = MdnsConfig::default();
     config.multicast_addr = *IPV6_MDNS_MULTICAST_ADDRESS;
-    run_test(MdnsConfig::default()).await
+    run_test(config).await
 }
```

**Comment by MariaGolubev** [comment]
I haven't looked here for a long time. The work done surprised me pleasantly. I'm glad @vnermolaev  has looked into this issue. Thank you

**Comment by dvc94ch** [comment]
This issue was fixed in #2359

**Comment by thomaseizinger** [comment]
Let's close it then!
